File: backend/app/services/auto_trade_service.py
"""
Background auto-trade tick: replaces manual buy/sell entry with automatic execution
against live NSE prices. Runs on an APScheduler interval job (see app/main.py).
"""
from __future__ import annotations

import logging
from datetime import datetime, time as dt_time
from decimal import Decimal
from zoneinfo import ZoneInfo

from app.database import SessionLocal
from app.models.execute_row_model import ExecuteRow
from app.models.script_model import Script
from app.schemas.execute_schema import ExecuteRowCreate
from app.services import execute_service, trigger_lifecycle
from app.services.telegram_alerts import send_telegram_message

logger = logging.getLogger(__name__)

IST = ZoneInfo("Asia/Kolkata")
MARKET_OPEN = dt_time(9, 15)
MARKET_CLOSE = dt_time(15, 30)


def is_market_open_now() -> bool:
    now = datetime.now(IST)
    if now.weekday() >= 5:  # Saturday=5, Sunday=6
        return False
    return MARKET_OPEN <= now.time() <= MARKET_CLOSE


def _auto_sell(db, script: Script, row: ExecuteRow, price: Decimal) -> None:
    payload = ExecuteRowCreate(
        trade_date=row.trade_date,
        trade_time=row.trade_time,
        buy_price=row.buy_price,
        sell_price=price,
    )
    updated = execute_service.update_execute_row(db, row.id, payload, notify=False)
    if not updated or updated.sell_price is None or updated.sell_qty is None:
        return
    pnl = ((updated.sell_price - updated.buy_price) * updated.sell_qty).quantize(Decimal("0.01"))
    send_telegram_message(
        f"🤖 AUTO-SELL executed\n"
        f"Script: {script.name}\n"
        f"Price: {updated.sell_price}\n"
        f"Qty: {updated.sell_qty}\n"
        f"P&L: {pnl}"
    )


def _auto_buy(db, script: Script, price: Decimal) -> None:
    now = datetime.now(IST)
    payload = ExecuteRowCreate(
        trade_date=now.date(),
        trade_time=now.time().replace(microsecond=0),
        buy_price=price,
        sell_price=None,
    )
    try:
        row = execute_service.add_execute_row(db, script.id, payload, notify=False)
    except ValueError:
        return
    if not row:
        return
    send_telegram_message(
        f"🤖 AUTO-BUY executed\nScript: {script.name}\nPrice: {row.buy_price}\nQty: {row.buy_qty}"
    )


def _process_script(db, script: Script) -> None:
    rows, _state, current, buy_t, _sell_t, _pairs = trigger_lifecycle.prepare_script_triggers(db, script)
    if current is None:
        return

    open_rows = sorted((r for r in rows if r.sell_price is None), key=lambda r: r.up_value)
    sold_any = False
    for row in open_rows:
        if current >= row.up_value:
            _auto_sell(db, script, row, current)
            sold_any = True

    if sold_any:
        rows, _state, current, buy_t, _sell_t, _pairs = trigger_lifecycle.prepare_script_triggers(db, script)
        if current is None:
            return
        open_rows = [r for r in rows if r.sell_price is None]

    # Don't open a new position once invest_slots concurrent positions are already held —
    # see execute_service.add_execute_row for the matching guard on the row-creation side.
    if len(open_rows) >= int(script.invest_slots):
        return

    if not rows:
        # Fresh script, never traded: enter immediately at the current price instead of
        # waiting for a dip. This also creates that row's own buy/sell targets (up_value/
        # down_value), which then drive all subsequent auto-buy/auto-sell decisions.
        _auto_buy(db, script, current)
    elif buy_t is not None and current <= buy_t:
        _auto_buy(db, script, current)


def run_tick() -> None:
    if not is_market_open_now():
        return
    db = SessionLocal()
    try:
        scripts = db.query(Script).filter(Script.active.is_(True)).all()
        for script in scripts:
            try:
                _process_script(db, script)
            except Exception:
                logger.exception("Auto-trade tick failed for script %s (%s)", script.id, script.name)
    finally:
        db.close()

File: backend/supabase/functions/_shared/autoTradeService.ts
import { d } from "./decimal.ts";
import { addExecuteRow, updateExecuteRow, NoSlotAvailableError, ZeroQtyError, type ExecuteRow } from "./executeService.ts";
import { sendTelegramMessage } from "./telegramAlerts.ts";
import * as tl from "./triggerLifecycle.ts";
import { indiaTimeStr, indiaWeekday } from "./wallclock.ts";
import type { Script } from "./scriptService.ts";
import { sql } from "./db.ts";

const MARKET_OPEN = "09:15:00";
const MARKET_CLOSE = "15:30:00";

export function isMarketOpenNow(now: Date = new Date()): boolean {
  if (indiaWeekday(now) >= 5) return false; // Saturday=5, Sunday=6
  const t = indiaTimeStr(now);
  return t >= MARKET_OPEN && t <= MARKET_CLOSE;
}

async function autoSell(script: Script, row: ExecuteRow, price: number): Promise<void> {
  const updated = await updateExecuteRow(
    row.id,
    { trade_date: row.trade_date, trade_time: row.trade_time, buy_price: Number(row.buy_price), sell_price: price },
    false,
  );
  if (!updated || updated.sell_price === null || updated.sell_qty === null) return;
  const pnl = d(updated.sell_price).minus(d(updated.buy_price)).times(d(updated.sell_qty, 6)).toDecimalPlaces(2);
  await sendTelegramMessage(
    `🤖 AUTO-SELL executed\nScript: ${script.name}\nPrice: ${updated.sell_price}\nQty: ${updated.sell_qty}\nP&L: ${pnl}`,
  );
}

async function autoBuy(script: Script, price: number): Promise<void> {
  const now = new Date();
  const [y, m, dd] = new Intl.DateTimeFormat("en-CA", { timeZone: "Asia/Kolkata", year: "numeric", month: "2-digit", day: "2-digit" })
    .format(now)
    .split("-");
  const tradeDate = `${y}-${m}-${dd}`;
  const tradeTime = indiaTimeStr(now);
  let row: ExecuteRow | null;
  try {
    row = await addExecuteRow(script.id, { trade_date: tradeDate, trade_time: tradeTime, buy_price: price, sell_price: null }, false);
  } catch (e) {
    if (e instanceof NoSlotAvailableError || e instanceof ZeroQtyError) return;
    throw e;
  }
  if (!row) return;
  await sendTelegramMessage(`🤖 AUTO-BUY executed\nScript: ${script.name}\nPrice: ${row.buy_price}\nQty: ${row.buy_qty}`);
}

async function processScript(script: Script): Promise<void> {
  let { rows, current, buyTrigger } = await tl.prepareScriptTriggers(script);
  if (current === null) return;

  let openRows = rows.filter((r) => r.sell_price === null).sort((a, b) => (d(a.up_value).minus(d(b.up_value))).toNumber());
  let soldAny = false;
  for (const row of openRows) {
    if (current.gte(d(row.up_value))) {
      await autoSell(script, row, current.toNumber());
      soldAny = true;
    }
  }

  if (soldAny) {
    const refreshed = await tl.prepareScriptTriggers(script);
    rows = refreshed.rows;
    current = refreshed.current;
    buyTrigger = refreshed.buyTrigger;
    if (current === null) return;
    openRows = rows.filter((r) => r.sell_price === null);
  }

  // Don't open a new position once invest_slots concurrent positions are already held.
  if (openRows.length >= script.invest_slots) return;

  if (rows.length === 0) {
    // Fresh script, never traded: enter immediately at the current price instead of waiting
    // for a dip. This also creates that row's own buy/sell targets (up_value/down_value),
    // which then drive all subsequent auto-buy/auto-sell decisions.
    await autoBuy(script, current.toNumber());
  } else if (buyTrigger !== null && current.lte(buyTrigger)) {
    await autoBuy(script, current.toNumber());
  }
}

export async function runTick(): Promise<void> {
  if (!isMarketOpenNow()) return;
  const scripts = await sql<Script[]>`select * from scripts where active = true`;
  for (const script of scripts) {
    try {
      await processScript(script);
    } catch (e) {
      console.error(`Auto-trade tick failed for script ${script.id} (${script.name}):`, e);
    }
  }
}
